### DISCUSSION TITLE
Physx backend

### DISCUSSION THREAD

**Discussion by woodbyte** [godotengine/godot-proposals#4044] [3D]
Would be good to have Physx available as an optional backend while Godot's own 3D physics module is still maturing.

This is doubly so for the upcoming Godot 4 release, which will no doubt attract the attention of many new 3D game developers but no longer features an alternative to fallback to as a way of avoiding bugs or for added performance.

**Comment by Sposito** [comment]
Also having latest physx in would be extremely valuable for folks like me using game engines for machine learning and robotics. Right now I am tied to unreal engine :/

**Reply by woodbyte** [reply]
> Also having latest physx in would be extremely valuable for folks like me using game engines for machine learning and robotics. Right now I am tied to unreal engine :/

I think Unity also added very precise joint simulation a few years back, specifically for robotics and using Physx, same as UE.

As for Godot, it would require not only a Physx backend, but also a bit of a revamp to joint nodes. Not enough is exposed to the user right now for such use cases.

**Reply by goatchurchprime** [reply]
Regarding accurate joint simulations (constraints), there's a proper mathematical way of doing it which doesn't require you to run the simulation at a very tiny timestep.  

This link seems to contrast the two approaches of "Spring Physics" and "PGS Physics" (doing the constraints properly): https://roblox.fandom.com/wiki/Physics_engine

Here's some slides outlining it: https://www.gdcvault.com/play/1026684/Math-In-Game-Development-Summit

Am I right in assuming that Godot is more of a spring physics engine, so can never handle these fixed constraints like joints properly?  (ie they are always done as tiny strong springs between rigid bodies?)

**Comment by eon-s** [comment]
On Godot 3 some users added things like Flex and Box2D, 4 is supposed to make it easier to add custom physics engines so there are chances to get these as plugins, there is no documentation about the process or how to use GDExtension in the case it is suitable for these applications.

**Comment by Calinou** [comment]
> This is doubly so for the upcoming Godot 4 release, which will no doubt attract the attention of many new 3D game developers but no longer features an alternative to fallback to as a way of avoiding bugs or for added performance.

There are no plans to add more physics engines as built-in functionality. This is partly to ensure the binary size remains low, but also to avoid issues with undermaintained physics engine integrations (as happened with Bullet).

Support for plugging in additional physics engines with GDExtension is planned, but it won't be finalized for 4.0 due to time constraints.[^1] Therefore, only GodotPhysics will be available in 4.0. On the bright side, this will surely channel and motivate people to fix bugs in GodotPhysics more quickly :slightly_smiling_face: 

[^1]: Using a [statically compiled C++ module](https://docs.godotengine.org/en/stable/development/cpp/custom_modules_in_cpp.html) will remain possible though, similar to https://github.com/Demindiro/godot_rapier3d for Godot 3.x.

**Reply by woodbyte** [reply]
Thanks for the update.

> On the bright side, this will surely channel and motivate people to fix bugs in GodotPhysics more quickly 🙂

Hah, that's certainly true. With an alternative in place there would be far less pressure to do so. The other side of that is, of course, that developers won't have an established library to compare their efforts against. But I suppose that's true of everything else in the engine. You compare with the previous iteration and that's good enough. 🙂

**Comment by UserProfileAccountName** [comment]
The PhysX 5.1 SDK is now available as open source and also includes Flow for fluid, fire and smoke simulation, and Blast for destruction

https://developer.nvidia.com/blog/open-source-simulation-expands-with-nvidia-physx-5-release/

https://github.com/NVIDIA-Omniverse/PhysX

**Comment by DrCanvas** [comment]
I don't know about full SDK integration. But we definitely need something solid for real-time cloth simulation.
Something like NVIDIA's NvCloth or even better. 
[https://gameworksdocs.nvidia.com/NvCloth/1.1/index.html](url)

**Reply by Calinou** [reply]
[GameWorks is under a proprietary license](https://github.com/NVIDIAGameWorks/NvCloth/blob/master/NvCloth/license.txt), so it can't be integrated in core Godot.

**Reply by UserProfileAccountName** [reply]
I think it's possible, O3DE has NvCloth as a Gem (plugin)

https://github.com/o3de/o3de/tree/development/Gems/NvCloth

https://www.o3de.org/docs/user-guide/gems/reference/physics/nvidia/nvidia-cloth/

https://www.o3de.org/docs/user-guide/interactivity/physics/nvidia-cloth/

**Reply by Calinou** [reply]
> I think it's possible, O3DE has NvCloth as a Gem (plugin)

Open source engines can integrate proprietary libraries, but they're not really open source anymore if they have a hard dependency on a proprietary library.

**Reply by UserProfileAccountName** [reply]
Based on a quick look at some files comparing the O3DE and the Nvidia repository, it appears that O3DE has modified NvCloth 

The plugin files also have this copyright notice

> [ Copyright (c) Contributors to the Open 3D Engine Project. 
For complete copyright and license terms please see the LICENSE at the root of this distribution.
SPDX-License-Identifier: Apache-2.0 OR MIT](https://github.com/o3de/o3de/blob/a468ec9b39d61bfb6ec36ee1c5ed5dc4e2cf6d9c/Gems/NvCloth/Code/Include/NvCloth/ISolver.h#L2)

I think they got a permission from Nvidia to release as open source

**Reply by Calinou** [reply]
The NvCloth includes in that folder you linked are only header files. It doesn't include the entire library (i.e. including `.cpp` files).

> I think they got a permission from Nvidia to release as open source

That's not how it works, especially with a company that has a large, structured legal department like NVIDIA :slightly_smiling_face:

**Reply by UserProfileAccountName** [reply]
Source files:
https://github.com/o3de/o3de/tree/main/Gems/NvCloth/Code/Source

**Reply by UserProfileAccountName** [reply]
> That's not how it works, especially with a company that has a large, structured legal department like NVIDIA 🙂

Considering they have the support of many companies (especially Amazon, the main contributor), I don't think it's impossible

https://www.o3de.org/
![1](https://user-images.githubusercontent.com/37518637/200957700-1f2bb201-12ed-4b21-b6cd-83ccfc133396.png)

**Comment by DiegoSynth** [comment]
I come 2 years later to very much support the Cloth physics request.
I've tried the impossible with SoftBody, figured out easier ways to deal with the pin points, set it all up... but the physics just won't work for scenarios involving constant collisions, such as skirts or garment that are in contact with the body.

I know it's not a completely fundamental feature, but nowadays it's quite standard and would make Godot show off.
Does anyone know if this is currently under consideration?

**Reply by Calinou** [reply]
>  Does anyone know if this is currently under consideration?

Considering we don't have a lot of physics contributors, not really. Right now, efforts are more likely to be put on [first-class Jolt integration](https://github.com/godotengine/godot-proposals/issues/7308) which are more essential for existing projects.

**Comment by uno1982** [comment]
https://github.com/godotengine/godot-proposals/discussions/15471

Just figured I’d link the discussion… after physx 5.10 open sourced a few months ago with all of the subsequent flex methods folded in and having gpu acceleration opened up to non cuda cards I simply couldn’t resist … the readme documents it’s a opt in build param and completely stand alone module … ci test currently passes 100% 

just finished liquid system today. Cloth, wind, rigid bodies traces etc are all in. Hope this finds you well 🫶

https://youtu.be/2fU7jvWatHI?is=LAJPaAUx962C8r25

**Comment by TOWUK** [comment]
Maybe I’m wrong, but hasn’t Nvidia stopped updating PhysX for almost 7 years now?
<img width="1159" height="568" alt="image" src="https://github.com/user-attachments/assets/996b311f-25e2-4f7d-bec9-cdeff75b49da" />

**Reply by Calinou** [reply]
This is the legacy PhysX for GPU-based physics additions in old games (< 2015), not the modern open source PhysX that's used as a physics engine.

**Comment by uno1982** [comment]
https://github.com/godotengine/godot-proposals/discussions/15471
WIP Fully physx5 simulated tanks .. motorcycles and vehicles done ✔️ 

https://github.com/user-attachments/assets/622c6435-47e0-4ba7-b928-16d6b2492ae3
